Tests/base/NSString/test09.m

```diff
@@ -3,55 +3,83 @@
+
 
   urlDecodedString = @"\1";
   urlEncodedString = @"%01";
   allowedCharacterSet = [NSCharacterSet alphanumericCharacterSet];
-  PASS(testUrlCharacterSetEncoding(urlDecodedString, urlEncodedString, allowedCharacterSet), "alphanumericCharacterSet");  
+  testEncodingString = [urlDecodedString stringByAddingPercentEncodingWithAllowedCharacters:allowedCharacterSet];
+  PASS_EQUAL(testEncodingString, urlEncodedString, "Percent-encoding one byte string");
+  PASS_EQUAL( [urlEncodedString stringByRemovingPercentEncoding], urlDecodedString, "Percent-encoded string decoding 05");
+
+
+  urlDecodedString = @"£";
+  urlEncodedString = @"%C2%A3";
+  allowedCharacterSet = [NSCharacterSet alphanumericCharacterSet];
+  testEncodingString = [urlDecodedString stringByAddingPercentEncodingWithAllowedCharacters:allowedCharacterSet];
+  PASS_EQUAL( testEncodingString, urlEncodedString, "Percent-encoding two byte character");
+  PASS_EQUAL([urlEncodedString stringByRemovingPercentEncoding], urlDecodedString, "Two-byte character decoding 06");
+
+
+  //check full string encoding and decoding
+  NSString* asIsString = @"0123456789 AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz _-~`!#$&'()*+,/:;=?@[]™…©®£ƒ‰¥§";
+  urlDecodedString = @"0123456789 AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz _-~`!#$%&'()*+,/:;=?@[]™…©®£ƒ‰¥§";
+  urlEncodedString  = @"%30%31%32%33%34%35%36%37%38%39%20%41%61%42%62%43%63%44%64%45%65%46%66%47%67%48%68%49%69%4A%6A%4B%6B%4C%6C%4D%6D%4E%6E%4F%6F%50%70%51%71%52%72%53%73%54%74%55%75%56%76%57%77%58%78%59%79%5A%7A%20%5F%2D%7E%60%21%23%24%25%26%27%28%29%2A%2B%2C%2F%3A%3B%3D%3F%40%5B%5D%E2%84%A2%E2%80%A6%C2%A9%C2%AE%C2%A3%C6%92%E2%80%B0%C2%A5%C2%A7";
+  allowedCharacterSet = [[NSCharacterSet characterSetWithCharactersInString:urlDecodedString] invertedSet];
+  testEncodingString = [urlDecodedString stringByAddingPercentEncodingWithAllowedCharacters:allowedCharacterSet];
+  PASS_EQUAL(testEncodingString, urlEncodedString, "Percent-encoding all characters in the string");
+  PASS_EQUAL([urlEncodedString stringByRemovingPercentEncoding], urlDecodedString, "Percent-encoded string decoding 07");
+
+
+  //check decoding of string with an unencoded part at the beginning
+  urlDecodedString = [asIsString stringByAppendingString:urlDecodedString];
+  urlEncodedString = [asIsString stringByAppendingString:urlEncodedString];
+  PASS_EQUAL([urlEncodedString stringByRemovingPercentEncoding], urlDecodedString, "Percent-encoded string decoding 08");
+
+
+  //check decoding of string with the encoded part in the middle
+  urlDecodedString = [urlDecodedString stringByAppendingString:asIsString];
+  urlEncodedString = [urlEncodedString stringByAppendingString:asIsString];
+  PASS_EQUAL([urlEncodedString stringByRemovingPercentEncoding], urlDecodedString, "Percent-encoded string decoding 09");
+
 
-  urlDecodedString = @"All alphabetic characters should be encoded. Symbols should not be: !@#$%^&*()_+-=";
```

Comment:
Why did you remove this special test case?

Reply:
This case covered by more extensive tests above